### DISCUSSION TITLE
Using `model_validate_json` on a model with `alias/alias_generator` and `populate_by_item=True` to only parse json by aliases.

### DISCUSSION THREAD

**Discussion by notypecheck** [pydantic/pydantic#9509] [Question]
I have to parse some json files and return them from API. Returned data should have the same format/naming as the initial files, the issue I'm facing right now is that it's possible for me to mistype a field alias, for example:

```json
{"a": 42}
```
```py
class Model(BaseModel):
    model_config = ConfigDict(populate_by_name=True)
    
    a: int = pydantic.Field(alias="Alias")
```
and this would return `{"Alias": 42}` when serializing it. Preferably I'd want want to catch those kind of errors.

**Comment by AlessandroMiola** [comment]
Hi @ThirVondukr,
not sure whether I completely get your point (sorry if I don't), but I'll try to answer anyway. 

If you mean that you're able to _deserialize_/_validate_ data by _field_name_ rather than by _alias_ only, that's due to [`populate_by_name=True`](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.populate_by_name) which allows you to possibly **deserialize** data **by field_name** (while always leaving you the (default) possibility of deserializing data by alias).

    from pydantic import BaseModel, ConfigDict, Field, ValidationError

    json_data = '{"a" : 42}'
    json_data_alias = '{"Alias" : 42}'
 
    class Model(BaseModel):
        model_config = ConfigDict(populate_by_name=True)
        a: int = Field(alias="Alias")

    try:
        m1 = Model.model_validate_json(json_data)
        m2 = Model.model_validate_json(json_data_alias)
    except ValidationError as ex:
        print(ex)

With such setting (`populate_by_name=True`) both `Model.model_validate_json(json_data)` and `Model.model_validate_json(json_data_alias)` would work. On the other hand, if you set `populate_by_name=False` (default) and you do have an alias specified, `Model.model_validate_json(json_data)` would raise a `ValidationError`.

If, instead, you mean that you would like data to be _serialized_ by _field_name_ and not by _alias_, that's the default. You do have the possibility of customizing serialization (so that serialized data will appear with aliases) by specifying [`by_alias=True`](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_dump) in `model_dump()`; it makes you **serialize** data **by alias**.

    try:
        m1 = Model.model_validate_json(json_data)
        m2 = Model.model_validate_json(json_data_alias)
        print(m1.model_dump(by_alias=True))    # --> {'Alias': 42}
        print(m1.model_dump())    # --> {'a': 42}
    except ValidationError as ex:
        print(ex)

As per my knowledge, here's a sort of recap of how things do work. 

**Default behaviours**:

- _(plain) aliases_: used for _deserialization_
- _field names_: used for _serialization_, model `repr`_esentation_ and for specifying _class attributes_

**(Main) Custom behaviours**:

- allow _deserialization_ by _field_name_: define a model level configuration that specifies [`populate_by_name=True`](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.populate_by_name)
- allow _serialization_ by _alias_: use `by_alias=True` in calls to [`model_dump()`](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_dump)
- allow _deserialization_ by [_validation alias_](https://docs.pydantic.dev/latest/concepts/alias/): pass `validation_alias` to `Field`; it overrides the `alias` specification (if present) when _deserializing_ data
- allow _serialization_ by [_serialization alias_](https://docs.pydantic.dev/latest/concepts/alias/): pass `serialization_alias` to `Field`; it overrides the `alias` specification (if present) when _serializing_ data (for instance when calling `model_dump(by_alias=True)`)

**Reply by notypecheck** [reply]
Hey, I use `populate_by_name` to be able to pass field names into it's constructor/init when instantiating my model from code. Thing I want is to always deserialize my models by alias 🤔, even if `populate_by_name` is present, perhaps there's an alternative for that config option?

**Reply by AlessandroMiola** [reply]
Hi @ThirVondukr,
I see your point now. I've just started exploring pydantic more thoroughly, so I'd leave more experienced people propose a better solution.

This said, I was curious about it and I tried experimenting a bit. As far as I know and I could experiment there's no configuration out of the box that would let you achieve such behaviour. 
One hack might be the following, namely overriding `model_validate` (the way I found maps `field_names` to the corresponding `aliases` and passes the `aliases` to `BaseModel.model_validate` as further objects to validate, but I suspect wiser solutions can be found 😝):

    data = {"a": 42}
    data_alias = {"alias": 42}
    data_random = {"b": 42}

    class Model(BaseModel):
        model_config = ConfigDict(populate_by_name=True)
        a: int = Field(alias="alias")

        @classmethod
        def model_validate(cls, obj):
            obj_alias = {cls.model_fields[key].alias: value for key, value in obj.items() if key in cls.model_fields}
            return super().model_validate(obj, **obj_alias)

    try:
        m1 = Model(a=42)   # works because of `populate_by_name=True`
        m2 = Model(alias=42)   # works (default behaviour)

        Model.model_validate(data)   # TypeError as `model_validate` ends up taking an unexpected keyword argument `alias`
        Model.model_validate(data_random)   # ValidationError raised by pydantic as a consequence of passing an argument which is neither the `field_name` nor the `alias`
        Model.model_validate(data_alias)   # works because `obj_alias` ends up being an _empty dict_ and `obj` has `alias` as the key
    except ValidationError as e:
        print(e)

For sure not the best (if not bad at all, let me know what you think...), because you end up catching an error which does not exploits the inherent `pydantic` validation capabilities (I mean, you end up with `pydantic` not raising a `ValidationError`, but a `TypeError` as a consequence of passing an unexpected keyword argument to `model_validate`).

I also thought about the possibility of exploiting a `BeforeValidator` at the _model level_, but this would also affect the possibility of instantiating fields by field_name while possibly allowing to you call `model_validate()` with the alias only.
